libopenage/gamestate/api/activity.cpp

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023-2023 the openage authors. See copying.md for legal info.
+
+#include "activity.h"
+
+#include "gamestate/api/definitions.h"
+
+
+namespace openage::gamestate::api {
+
+bool APIActivity::is_activity(const nyan::Object &obj) {
+	nyan::fqon_t immediate_parent = obj.get_parents()[0];
+	return immediate_parent == "engine.util.activity.Activity";
+}
+
+nyan::Object APIActivity::get_start(const nyan::Object &activity) {
+	auto obj_value = activity.get<nyan::ObjectValue>("Activity.start");
+
+	std::shared_ptr<nyan::View> db_view = activity.get_view();
+	return db_view->get_object(obj_value->get_name());
+}
+
+
+bool APIActivityNode::is_node(const nyan::Object &obj) {
+	nyan::fqon_t immediate_parent = obj.get_parents()[0];
+	return immediate_parent == "engine.util.activity.node.Node";
+}
+
+activity::node_t APIActivityNode::get_type(const nyan::Object &node) {
+	nyan::fqon_t immediate_parent = node.get_parents()[0];
+	return ACTIVITY_NODE_DEFS.get(immediate_parent);
+}
+
+std::vector<nyan::Object> APIActivityNode::get_next(const nyan::Object &node) {
+	switch (APIActivityNode::get_type(node)) {
+	// 0 next nodes
+	case activity::node_t::END: {
+		return {};
+	}
+	// 1 next node
+	case activity::node_t::TASK_SYSTEM: {
+		auto next = node.get<nyan::ObjectValue>("Ability.next");
+		std::shared_ptr<nyan::View> db_view = node.get_view();
+		return {db_view->get_object(next->get_name())};
+	}
+	case activity::node_t::START: {
+		auto next = node.get<nyan::ObjectValue>("Start.next");
+		std::shared_ptr<nyan::View> db_view = node.get_view();
+		return {db_view->get_object(next->get_name())};
+	}
+	// 1+ next nodes
+	case activity::node_t::XOR_GATE: {
```

Comment:
shouldn't this be part of the xorgate node implementation instead of this central place? might be a suitable application of a std variant and stdvisit.

Reply:
I don't want to integrate nyan too deep into the gamestate. That's why these helper classes exist as a kind of layer between nyan and openage. I think this also makes it easier to maintain the API when we change it.

Reply:
but isn't the xorgate node class the abstraction layer? so instead of having specific snippets here, having them more local to the specific code using it seems more local.

Reply:
I don't think this would make more sense though. The function here is only used to traverse the activity graph defined in the modpack, so it doesn't have much to do with activities at runtime. It's only use is to create the runtime activity from the nyan definition. XorGate and the other nodes have their own internal `next` function for traversing the runtime graph.